File: solutions/cpp/largest-series-product/largest_series_product.cpp
#include "largest_series_product.h"

#include <stdexcept>
#include <algorithm>

namespace largest_series_product {

    auto largest_product(const std::string &s, int l) -> int
    {
        if (l < 1 || l > static_cast<int>(s.length()))
            throw std::domain_error("");

        std::vector<std::string> slices;

        for (std::size_t i = 0; i <= s.length() - l; i++)
        {
            slices.push_back(s.substr(i, l));
        }

        for (const auto &slice : slices)
        {
            bool is_valid =
                std::all_of (
                    slice.begin(),
                    slice.end(),
                    [](unsigned char ch) { return std::isdigit(ch); }
                );

            if (!is_valid)
                throw std::domain_error("");
        }

        std::vector<int> products;

        for (const auto &slice : slices)
        {
            int product = 1;

            for (const auto &c : slice)
            {
                product *= c - '0';
            }

            products.push_back(product);
        }

        return *std::max_element(products.begin(), products.end());
    }

}  // namespace largest_series_product

File: solutions/csharp/largest-series-product/LargestSeriesProduct.cs
public static class LargestSeriesProduct
{
    public static long GetLargestProduct(string digits, int span)
    {
        if (span < 0 || span > digits.Length)
            throw new ArgumentException("");

        var slices = new string[digits.Length - span + 1];

        for (int i = 0; i < slices.Length; i++)
            slices[i] = digits.Substring(i, span);

        var products = new long[slices.Length];

        for (int i = 0; i < slices.Length; i++)
        {
            products[i] = 1;
            foreach (char c in slices[i])
            {
                if (!char.IsDigit(c))
                    throw new ArgumentException("");
                products[i] *= c - '0';
            }
        }

        return products.Max();
    }
}